core/common/src/test/java/alluxio/resource/PooledResourceTest.java

```diff
@@ -0,0 +1,122 @@
+  @Test
+  public void releaseOnClose() {
+    TestPool pool = new TestPool(5);
+    try (PooledResource<Resource> i = new PooledResource<>(pool.acquire(), pool)) {
+      assertFalse(i.get().isClosed());
+      assertEquals(0, pool.getAvailableCount());
+      assertEquals(1, pool.getLeakCount());
+    }
+    assertEquals(1, pool.getAvailableCount());
+    assertEquals(0, pool.getLeakCount());
+  }
+
+  @Test
+  public void leakDoesNotPreventPoolFromBeingGCed() throws Exception {
+    long gcTimeout = 1000;
+    TestPool pool = new TestPool(5);
+    ReferenceQueue<TestPool> referenceQueue = new ReferenceQueue<>();
+    PhantomReference<TestPool> poolRef = new PhantomReference<>(pool, referenceQueue);
+    PooledResource<Resource> notLeaked = new PooledResource<>(pool.acquire(), pool);
+    PooledResource<Resource> leaked = new PooledResource<>(pool.acquire(), pool);
+    assertFalse(notLeaked.get().isClosed());
+    assertFalse(leaked.get().isClosed());
+    assertEquals(0, pool.getAvailableCount());
+    assertEquals(2, pool.getLeakCount());
+
+    notLeaked.close();
+    // we don't assert on the closed-ness of the resource wrapped in notLeaked here,
+    // since it's released to the pool, and the pool will decide when to close it
+    assertEquals(1, pool.getLeakCount());
+    assertEquals(1, pool.getAvailableCount());
+
+    // GC the pool before toLeak is released
+    pool = null;
+    System.gc();
+
+    // check the pool has really been GCed
+    Reference<? extends TestPool> enqueued = referenceQueue.remove(gcTimeout);
+    assertNotNull("ref is null, pool is likely not GCed before timeout", enqueued);
+    assertEquals(poolRef, enqueued);
+    enqueued.clear();
+
+    // we should still be able to close this without getting an exception
+    Resource leakedResource = leaked.get();
+    leaked.close();
+    assertTrue(leakedResource.isClosed());
+  }
+
+  static class Resource implements AutoCloseable {
+    private boolean mIsClosed = false;
+
+    @Override
+    public void close() throws Exception {
+      mIsClosed = true;
+    }
+
+    public boolean isClosed() {
+      return mIsClosed;
+    }
+  }
+
+  static class TestPool extends ResourcePool<Resource> {
+    TestPool(int capacity) {
+      super(capacity, new ConcurrentLinkedQueue<>());
+    }
+
+    @Override
+    public void close() throws IOException {
+      // do nothing
+    }
+
+    @Override
+    public Resource createNewResource() {
+      return new Resource();
+    }
+
+    /**
+     * Gets the number of resources that are immediately available in the pool, without having to
+     * call {@link #createNewResource()}.
+     *
+     * @return number of resources
+     */
+    public int getAvailableCount() {
+      return mResources.size();
+    }
+
+    /**
+     * Gets the number of resources that has been acquired from this pooled but not released yet.
+     *
+     * @return the number of resources that is considered leaked
+     */
```

Comment:
The name of this function is frightening.
Maybe we can rename it as:
`getUsedCount()`
?

Reply:
This is test code, and the purpose of the test is to ensure the leaked resource does not cause the pool to be leaked, too. So I say let's make the method name explicit.